File: week4/이다혜/양궁대회.java
public class 양궁대회 {

    public int[] lion;
    public int max = -1;
    public int [] res = {-1, };
    public int[] solution(int n, int[] info) {

        lion = new int[11];

        recursiveSearch(n, info, 0);

        return res;
    }

    public void recursiveSearch(int n, int [] info, int cnt) {

        if(cnt == n) {
            int lionScore = 0, apeachScore = 0;
            for(int i = 0; i <= 10; i++) {
                if(info[i] > 0 || lion[i] > 0) {
                    if(lion[i] > info[i]) {
                        lionScore += 10 - i;
                    } else apeachScore += 10 - i;
                }
            }

            if(lionScore > apeachScore) {
                if(lionScore - apeachScore >= max) {
                    res = lion.clone();
                    max = lionScore - apeachScore;
                }
            }

            return ;
        }

        for(int i = 0; i <= 10 && lion[i] <= info[i]; i++) {
            lion[i]++;
            recursiveSearch(n, info, cnt+1);
            lion[i]--;
        }
    }
}

File: week4/한지희/양궁대회.kt
class ArcheryGames {
    private lateinit var apeach: IntArray
    private lateinit var lion: IntArray
    private var answer = intArrayOf(-1)
    private var max = Int.MIN_VALUE
    private var remainArrow = 0

    fun solution(n: Int, info: IntArray): IntArray {
        apeach = info
        remainArrow = n
        lion = IntArray(11)
        dfs(0)
        return answer
    }

    private fun dfs(depth: Int) {
        if (depth == remainArrow) {
            val difScore = getLionApeachDiffScore()
            if (difScore > 0 && difScore >= max) {
                this.answer = lion.clone()
                max = difScore
            }
            return
        }

        for (i in 0..10) {
            if (lion[i] > apeach[i]) break
            lion[i]++
            dfs(depth + 1)
            lion[i]--
        }
    }

    private fun getLionApeachDiffScore(): Int {
        var apeachTotal = 0
        var lionTotal = 0
        for (i in 0..10) {
            if (apeach[i] != 0 || lion[i] != 0) {
                if (apeach[i] < lion[i]) lionTotal += 10 - i
                else apeachTotal += 10 - i
            }
        }
        return lionTotal - apeachTotal
    }
}